File: web/src/lib/description-labels.ts
/**
 * The ranges of a description the drawing does not read.
 *
 * The rule is enforced on the server (`server/src/inku_server/description_labels.py`);
 * this is the editor's copy, and it exists only so the author can see which
 * ranges will be dropped.  Because the rule now lives in two languages, both
 * read the same corpus: `server/tests/data/description-label-cases.json`.
 * Change one side and description-labels.test.ts turns red.
 */

export type LabelSpanKind = 'number' | 'comment';

export type LabelSpan = {
	start: number;
	end: number;
	kind: LabelSpanKind;
};

// Digits of either width, then a separator an author actually types.  The
// ideographic space counts as one, because "１　花" is how a numbered line is
// written in Japanese; a plain space does not, or "3 本の線" would lose its 3.
const LEADING_NUMBER = /^[ \t]*[0-9０-９]+[.．、)）:：　][ \t　]*/;

// Brackets of either width, closed on the line they were opened on.  An
// unclosed "[" is description: it would otherwise swallow the rest of the text.
const COMMENT = /\[[^[\]\n]*\]|［[^［］\n]*］/g;

/** Every range the drawing does not read, in order of appearance. */
export function excludedSpans(text: string): LabelSpan[] {
	if (!text) return [];
	const spans: LabelSpan[] = [];
	let offset = 0;
	// Keep the line breaks: the offsets are into the original string.
	for (const line of text.split('\n')) {
		const number = LEADING_NUMBER.exec(line);
		if (number) spans.push({ start: offset, end: offset + number[0].length, kind: 'number' });
		COMMENT.lastIndex = 0;
		let comment: RegExpExecArray | null;
		while ((comment = COMMENT.exec(line)) !== null) {
			spans.push({ start: offset + comment.index, end: offset + comment.index + comment[0].length, kind: 'comment' });
		}
		offset += line.length + 1; // + the "\n" that split() removed
	}
	spans.sort((a, b) => a.start - b.start);
	return spans;
}

export type LabelSegment = {
	text: string;
	kind: LabelSpanKind | null;
};

/**
 * The text split into what the drawing reads and what it does not, in order.
 * The editor paints a background behind every segment that has a kind.
 */
export function labelSegments(text: string): LabelSegment[] {
	const spans = excludedSpans(text);
	if (spans.length === 0) return text ? [{ text, kind: null }] : [];
	const segments: LabelSegment[] = [];
	let at = 0;
	for (const span of spans) {
		if (span.start > at) segments.push({ text: text.slice(at, span.start), kind: null });
		segments.push({ text: text.slice(span.start, span.end), kind: span.kind });
		at = span.end;
	}
	if (at < text.length) segments.push({ text: text.slice(at), kind: null });
	return segments;
}

/**
 * What the server will read.  The editor does not send this -- the server cuts
 * for every client -- but the character meter counts it, because the guide is
 * about the description, not about the author's numbering.
 */
export function pipelineDescription(text: string): string {
	if (!text) return text ?? '';
	const spans = excludedSpans(text);
	if (spans.length === 0) return text;
	const kept: string[] = [];
	let at = 0;
	for (const span of spans) {
		kept.push(text.slice(at, span.start));
		at = span.end;
	}
	kept.push(text.slice(at));
	return kept
		.join('')
		.split('\n')
		.map((line) => line.replace(/[ \t　]{2,}/g, ' ').replace(/^[ \t　]+|[ \t　]+$/g, ''))
		.join('\n');
}

File: server/src/inku_server/description_labels.py
"""What the author writes for themselves, and the drawing never reads.

Two things in a description are labels rather than description:

  - a **leading number** ("1. ", "０１．", "12) ") that orders lines in a batch;
  - a **comment** in brackets ("[疎  紀友則 / 古今和歌集（春下）]") naming a source.

Both are kept verbatim in the stored work -- they are the author's document --
and both are cut before the text reaches any layer, Stage 0.5 included.  The cut
happens here, once, so that every client gets the same rule: the web editor
greys these ranges out, but the web is not what enforces them.

The functions return spans as well as text so the editor can paint exactly what
the server will drop.
"""

from __future__ import annotations

import re
from typing import NamedTuple

# A numbering has to be a numbering, not a measurement: "3本の線" and "2026年"
# are description.  Digits (either width) followed by one of the separators an
# author actually types, and the ideographic space counts as one because "１　花"
# is how a numbered line is written in Japanese.
_LEADING_NUMBER = re.compile(r"^[ \t]*[0-9０-９]+[.．、)）:：　][ \t　]*")

# Brackets of both widths, closed on the line they were opened on.  An unclosed
# "[" is description: it would otherwise swallow the rest of the text.
_COMMENT = re.compile(r"\[[^\[\]\n]*\]|［[^［］\n]*］")


class Span(NamedTuple):
    """A half-open range of the original text that the drawing does not read."""

    start: int
    end: int
    kind: str  # "number" | "comment"


def excluded_spans(text: str) -> list[Span]:
    """Every range the drawing does not read, in order of appearance."""
    if not text:
        return []
    spans: list[Span] = []
    offset = 0
    for line in text.splitlines(keepends=True):
        number = _LEADING_NUMBER.match(line)
        if number:
            spans.append(Span(offset + number.start(), offset + number.end(), "number"))
        for comment in _COMMENT.finditer(line):
            spans.append(Span(offset + comment.start(), offset + comment.end(), "comment"))
        offset += len(line)
    spans.sort()
    return spans


def pipeline_description(text: str | None) -> str:
    """The description with its labels removed, as every layer should read it.

    Whitespace left behind by a removal is collapsed so that a comment taken out
    of the middle of a sentence does not leave a double space; line breaks are
    kept, because they are the author's.
    """
    if not text:
        return text or ""
    spans = excluded_spans(text)
    if not spans:
        return text
    kept: list[str] = []
    at = 0
    for start, end, _kind in spans:
        kept.append(text[at:start])
        at = end
    kept.append(text[at:])
    stripped = "".join(kept)
    # Collapse the gap a removal leaves, per line, without touching the breaks.
    lines = [re.sub(r"[ \t　]{2,}", " ", line).strip(" \t　") for line in stripped.split("\n")]
    return "\n".join(lines)
